File: android_claude/data/crypto/src/main/java/com/zarabandajose/transcripto/data/crypto/cipher/Base64Codec.kt
package com.zarabandajose.transcripto.data.crypto.cipher

import com.zarabandajose.transcripto.core.common.Result
import com.zarabandajose.transcripto.domain.model.CipherParams
import java.util.Base64

/**
 * Implementación del codec Base64.
 * Utiliza java.util.Base64 para encoding/decoding con soporte UTF-8.
 */
class Base64Codec {
    
    /**
     * Codifica texto a Base64, con soporte para salt.
     */
    fun encrypt(input: String, params: CipherParams): Result<String> {
        return try {
            val textToEncode = if (params.useSalt && params.salt.isNotEmpty()) {
                "${params.salt}:$input" // Preprend salt con separador
            } else {
                input
            }
            
            val encoded = Base64.getEncoder().encodeToString(textToEncode.toByteArray(Charsets.UTF_8))
            Result.Success(encoded)
        } catch (e: Exception) {
            Result.Error("Error al codificar Base64: ${e.message}", e)
        }
    }
    
    /**
     * Decodifica texto desde Base64, removiendo salt si está presente.
     */
    fun decrypt(input: String, params: CipherParams): Result<String> {
        return try {
            val decoded = String(Base64.getDecoder().decode(input), Charsets.UTF_8)
            
            val result = if (params.useSalt && params.salt.isNotEmpty()) {
                // Buscar y remover el salt del inicio
                val saltPrefix = "${params.salt}:"
                if (decoded.startsWith(saltPrefix)) {
                    decoded.substring(saltPrefix.length)
                } else {
                    // Si no encuentra el salt esperado, devolver el texto completo
                    decoded
                }
            } else {
                decoded
            }
            
            Result.Success(result)
        } catch (e: IllegalArgumentException) {
            Result.Error("Formato Base64 inválido", e)
        } catch (e: Exception) {
            Result.Error("Error al decodificar Base64: ${e.message}", e)
        }
    }
}

File: ios_claude/transcripto_calude/transcripto_calude/DataCrypto/Base64Codec/Base64Codec.swift
//
//  Base64Codec.swift
//  transcripto_calude
//
//  Created by JOSE ZARABANDA on 9/3/25.
//

import Foundation

// Error local para Base64
enum Base64Error: LocalizedError {
    case encryptionFailed(String)
    case invalidBase64
    case decryptionFailed(String)
    
    var errorDescription: String? {
        switch self {
        case .encryptionFailed(let message):
            return "Error de cifrado: \(message)"
        case .invalidBase64:
            return "El texto no es un Base64 válido"
        case .decryptionFailed(let message):
            return "Error de descifrado: \(message)"
        }
    }
}

/// Implementación del codec Base64
public class Base64Codec: CipherMethod {
    
    public init() {}
    
    public func encrypt(_ input: String, params: CipherParams) throws -> String {
        try InputValidator.validateOperation(text: input, params: params)
        
        var textToEncode = input
        
        // Aplicar salt antes de codificar si está presente
        if let salt = params.salt {
            textToEncode = CryptoUtils.applySaltToText(input, salt: salt, operation: .encrypt)
        }
        
        guard let data = textToEncode.data(using: .utf8) else {
            throw Base64Error.encryptionFailed("No se pudo convertir el texto a datos UTF-8")
        }
        
        return data.base64EncodedString()
    }
    
    public func decrypt(_ input: String, params: CipherParams) throws -> String {
        try InputValidator.validateOperation(text: input, params: params, isDecryption: true)
        
        guard let data = Data(base64Encoded: input) else {
            throw Base64Error.invalidBase64
        }
        
        guard let decodedText = String(data: data, encoding: .utf8) else {
            throw Base64Error.decryptionFailed("No se pudo convertir los datos a texto UTF-8")
        }
        
        var result = decodedText
        
        // Remover salt si está presente
        if let salt = params.salt {
            result = CryptoUtils.applySaltToText(decodedText, salt: salt, operation: .decrypt)
        }
        
        return result
    }
}
